File: PHP/src/NyzoStringMicropay.php
<?php
  /**
   * ref: https://github.com/n-y-z-o/nyzoVerifier/blob/master/src/main/java/co/nyzo/verifier/nyzoString/NyzoStringMicropay.java
   * version: 0.0.2
   */
  require_once("NyzoString.php");
  require_once("NyzoStringType.php");
  require_once("utils/Buffer.php");
  require_once("utils/Int64.php");
  
  final class NyzoStringMicropay implements NyzoString {

    private $receiverIdentifier;
    private $senderData;
    private $amount;
    private $timestamp;
    private $previousHashHeight;
    private $previousBlockHash;

    public function __construct(Buffer $receiverIdentifier, Buffer $senderData, Int64 $amount, Int64 $timestamp,
    Int64 $previousHashHeight, Buffer $previousBlockHash) {
      $senderData = $senderData->slice(0, 32);
      $this->receiverIdentifier = $receiverIdentifier;
      $this->senderData = $senderData;
      $this->amount = $amount;
      $this->timestamp = $timestamp;
      $this->previousHashHeight = $previousHashHeight;
      $this->previousBlockHash = $previousBlockHash;
    }

    public function getReceiverIdentifier(): Buffer {
      return $this->receiverIdentifier;
    }

    public function getSenderData(): Buffer {
      return $this->senderData;
    }

    public function getAmount(): Int64 {
      return $this->amount;
    }

    public function getTimestamp(): Int64 {
      return $this->timestamp;
    }

    public function getPreviousHashHeight(): Int64 {
      return $this->previousHashHeight;
    }

    public function getPreviousBlockHash(): Buffer {
      return $this->previousBlockHash;
    }

    public function getType(): NyzoStringType {
      return NyzoStringType::Micropay();
    }
    public function getBytes(): array {
      $bytes = new Buffer(32 + 1 + $this->senderData->size() + 8 + 8 + 8 + 32);
      $this->receiverIdentifier->copy($bytes, 0);
      $senderDataLength = $this->senderData->size();
      $bytes[32] = $senderDataLength;
      $this->senderData->copy($bytes, 33);
      $this->amount->copy($bytes, 33 + $senderDataLength);
      $this->timestamp->copy($bytes, 33 + $senderDataLength + 8);
      $this->previousHashHeight->copy($bytes, 33 + $senderDataLength + 8 + 8);
      $this->previousBlockHash->copy($bytes, 33 + $senderDataLength + 8 + 8 + 8);
      return $bytes->toArray();
    }

    public static function fromByteBuffer(Buffer $buffer): NyzoStringMicropay {
      $receiverBuffer = $buffer->slice(0, 32);
      $senderDataLength = min($buffer[32] & 0xff, 32);
      $dataBuffer = $buffer->slice(33, 33 + $senderDataLength);
      $amount = new Int64($buffer->slice(33 + $senderDataLength, 33 + $senderDataLength + 8));
      $timestamp = new Int64($buffer->slice(33 + $senderDataLength + 8, 33 + $senderDataLength + 8 + 8));
      $previousHashHeight = new Int64($buffer->slice(33 + $senderDataLength + 8 + 8, 33 + $senderDataLength + 8 + 8 + 8));
      $previousBlockHashBuffer = $buffer->slice(33 + $senderDataLength + 8 + 8 + 8, 33 + $senderDataLength + 8 + 8 + 8 + 32);
      return new NyzoStringMicropay($receiverBuffer, $dataBuffer, $amount, $timestamp, $previousHashHeight, $previousBlockHashBuffer);
    }

    public static function fromHex(string $receiverHexString, string $dataHexString, string $amountHex, string $timestampHex, string $previousHashHeightHex, string $previousBlockHashBufferHex): NyzoStringMicropay {
      $filteredString = substr(implode("", explode("-", $receiverHexString)), 0, 64);
      $receiverBuffer = Buffer::fromHex($filteredString);
      $dataBuffer = Buffer::fromHex($dataHexString);
      $amount = new Int64(Buffer::fromHex($amountHex));
      $timestamp = new Int64(Buffer::fromHex($timestampHex));
      $previousHashHeight = new Int64(Buffer::fromHex($previousHashHeightHex));
      $filteredHash = substr(implode("", explode("-", $previousBlockHashBufferHex)), 0, 64);
      $previousBlockHashBuffer = Buffer::fromHex($filteredHash);
      return new NyzoStringMicropay($receiverBuffer, $dataBuffer, $amount, $timestamp, $previousHashHeight, $previousBlockHashBuffer);
    }
  }
?>

File: Python/nyzostrings/nyzostrings/nyzostringmicropay.py
"""
Nyzo String for micropay
"""

from nyzostrings.nyzostring import NyzoString


__version__ = "0.0.1"


class NyzoStringMicropay(NyzoString):
    def __init__(
        self,
        receiver_identifier: bytes,
        sender_data: bytes,
        amount: int,
        timestamp: int,
        previous_hash_height: int,
        previous_block_hash: bytes,
    ) -> None:
        index = 0
        sender_data = sender_data[:32]
        sender_data_len = len(sender_data)
        bytes_content = bytearray(32 + 1 + sender_data_len + 8 + 8 + 8 + 32)
        bytes_content[0:32] = receiver_identifier
        index += 32
        bytes_content[index] = sender_data_len
        index += 1
        bytes_content[index : index + sender_data_len] = sender_data
        index += sender_data_len
        bytes_content[index : index + 8] = amount.to_bytes(8, byteorder="big")
        index += 8
        bytes_content[index : index + 8] = timestamp.to_bytes(8, byteorder="big")
        index += 8
        bytes_content[index : index + 8] = previous_hash_height.to_bytes(
            8, byteorder="big"
        )
        index += 8
        bytes_content[index : index + 32] = previous_block_hash
        super().__init__("pay_", bytes_content)
        self.receiver_identifier = receiver_identifier
        self.sender_data = sender_data
        self.amount = amount
        self.timestamp = timestamp
        self.previous_hash_height = previous_hash_height
        self.previous_block_hash = previous_block_hash

    @classmethod
    def from_bytes(cls, byte_buffer: bytes) -> "NyzoStringMicropay":
        buffer = memoryview(byte_buffer)
        index = 0
        receiver_buffer = buffer[index : index + 32]
        index += 32
        sender_data_length = min(buffer[32] & 0xFF, 32)
        index += 1
        data_buffer = buffer[index : index + sender_data_length]
        index += sender_data_length
        amount = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        timestamp = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        previous_hash_height = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        previous_block_hash_buffer = buffer[index : index + 32]
        return NyzoStringMicropay(
            receiver_buffer,
            data_buffer,
            amount,
            timestamp,
            previous_hash_height,
            previous_block_hash_buffer,
        )

    @classmethod
    def from_hex(
        cls,
        receiver_identifier_hex: str,
        sender_data_hex: str,
        amount_hex: str,
        timestamp_hex: str,
        previous_hash_height_hex: str,
        previous_block_hash_hex: str,
    ) -> "NyzoStringMicropay":
        filtered_string = receiver_identifier_hex.replace("-", "")[:64]
        receiver = bytes.fromhex(filtered_string)
        filtered_string = sender_data_hex.replace("-", "")[:64]
        data = bytes.fromhex(filtered_string)
        amount = int(amount_hex, 16)
        timestamp = int(timestamp_hex, 16)
        previous_hash_height = int(previous_hash_height_hex, 16)
        filtered_string = previous_block_hash_hex.replace("-", "")[:64]
        previous_block_hash = bytes.fromhex(filtered_string)
        return NyzoStringMicropay(
            receiver, data, amount, timestamp, previous_hash_height, previous_block_hash
        )
